doc/development/plugins.rst

```diff
@@ -1,397 +1,398 @@
+
+Devices may define their own transforms following the description in the ``transforms`` module,
+or can include in-built transforms such as:
+
+.. currentmodule:: pennylane.devices.preprocess
 .. autosummary::
+    :toctree: api
 
-    apply
+    decompose
+    validate_observables
+    validate_measurements
+    validate_device_wires
+    validate_multiprocessing_workers
+    validate_adjoint_trainable_params
+    no_sampling
 
-If the device is a statevector simulator (it can perform analytic computations when ``shots=None``)
-then it **must** also overwrite:
 
+.. currentmodule:: pennylane
 .. autosummary::
+    :toctree: api
 
-    analytic_probability
-
-The :class:`~.QubitDevice` class
-provides the following convenience methods that may be used by the plugin:
+    defer_measurements
+    dynamic_one_shot
+    transforms.broadcast_expand
+    transforms.split_non_commuting
+    transforms.transpile
+    transforms.diagonalize_measurements
+    transforms.split_to_single_terms
 
-.. autosummary::
-
-    active_wires
-    marginal_prob
 
-In addition, if your qubit device generates its own computational basis samples for measured wires
-after execution, you need to overwrite the following method:
+Wires
+-----
 
-.. autosummary::
+Devices can now either:
 
-    generate_samples
+1) Strictly use wires provided by the user on initialization ``device(name, wires=wires)``
+2) Infer the number and ordering of wires provided by the submitted circuit.
+3) Strictly require specific wire labels
 
-:meth:`~.QubitDevice.generate_samples` should return samples with shape ``(dev.shots, dev.num_wires)``.
-Furthermore, PennyLane uses the convention :math:`|q_0,q_1,\dots,q_{N-1}\rangle` where
-:math:`q_0` is the most significant bit.
+Option 2 allows workflows to change the number and labeling of wires over time, but sometimes users want
+to enfore a wire convention and labels. If a user does provide wires, the :meth:`~.devices.Device.preprocess` should
+validate that submitted circuits only have wires in the requested range.
 
-And thats it! The device has inherited :meth:`~.QubitDevice.expval`, :meth:`~.QubitDevice.var`,
-and :meth:`~.QubitDevice.sample` methods, that accepts an observable (or tensor product of
-observables) and returns the corresponding measurement statistic.
+>>> dev = qml.device('default.qubit', wires=1)
+>>> circuit = qml.tape.QuantumScript([qml.CNOT((0,1))], [qml.state()])
+>>> dev.preprocess()[0]((circuit, ))
+WireError: Cannot run circuit(s) of default.qubit as they contain wires not found on the device.
 
+PennyLane wires can be any hashable object, where wire labels are distinguished by their equality and hash.
+If working with successive integers (``0``, ``1``, ``2``, ...) is preferred internally,
+the :meth:`~.QuantumScript.map_to_standard_wires` method can be used inside of 
+the :meth:`~.devices.Device.execute` method.  The :class:`~.map_wires` transform can also 
```

Comment:
```suggestion
the :meth:`~.devices.Device.execute` method. The :class:`~.map_wires` transform can also 
```

Reply:
sorry... a little to hard to change such an ingrained habit...